src/main/java/htsjdk/samtools/CRAMBAIIndexer.java

```diff
@@ -115,56 +113,34 @@ public CRAMBAIIndexer(final OutputStream output, final SAMFileHeader fileHeader)
-                    final ContainerParser parser = new ContainerParser(indexBuilder.bamHeader);
-                    final Map<Integer, AlignmentSpan> refSet = parser.getReferences(container, validationStringency);
-                    final Slice fakeSlice = new Slice();
-                    slice.containerOffset = container.offset;
-                    slice.index = sliceIndex++;
-                    /**
-                     * Unmapped span must be processed after mapped spans:
-                     */
-                    AlignmentSpan unmappedSpan = refSet.remove(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
-                    for (final int refId : new TreeSet<>(refSet.keySet())) {
-                        final AlignmentSpan span = refSet.get(refId);
-                        fakeSlice.sequenceId = refId;
-                        fakeSlice.containerOffset = slice.containerOffset;
-                        fakeSlice.offset = slice.offset;
-                        fakeSlice.index = slice.index;
-
-                        fakeSlice.alignmentStart = span.getStart();
-                        fakeSlice.alignmentSpan = span.getSpan();
-                        fakeSlice.nofRecords = span.getCount();
-                        processSingleReferenceSlice(fakeSlice);
-                    }
-                    if (unmappedSpan != null) {
-                        final AlignmentSpan span = unmappedSpan;
-                        fakeSlice.sequenceId = SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX;
-                        fakeSlice.containerOffset = slice.containerOffset;
-                        fakeSlice.offset = slice.offset;
-                        fakeSlice.index = slice.index;
-
-                        fakeSlice.alignmentStart = SAMRecord.NO_ALIGNMENT_START;
-                        fakeSlice.alignmentSpan = 0;
-                        fakeSlice.nofRecords = span.getCount();
-                        processSingleReferenceSlice(fakeSlice);
-                    }
-                } else {
-                    processSingleReferenceSlice(slice);
+        for (final IndexableSlice slice : container.slices) {
+            if (slice.hasMultipleReferences()) {
+                final ContainerParser parser = new ContainerParser(indexBuilder.bamHeader);
+                final Map<Integer, AlignmentSpan> referenceAlignmentMap = parser.getReferences(container, validationStringency);
+
+                /**
+                 * Unmapped span must be processed after mapped spans:
+                 */
+                final AlignmentSpan unmappedSpan = referenceAlignmentMap.remove(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
+                final Set<Integer> mappedReferences = new TreeSet<>(referenceAlignmentMap.keySet());
+
+                for (final int refId : mappedReferences) {
+                    final AlignmentSpan span = referenceAlignmentMap.get(refId);
+                    processSingleReferenceSlice(slice.getBAIMetadata(refId, span, container.offset));

```

Comment:
instead of creating a `fakeSlice` we generate the BAI Metadata from a real `Slice` and a few parameters